mlflow/openai/__init__.py

```diff
@@ -576,22 +576,66 @@ def _has_content_and_role(d):
     return "content" in d and "role" in d
 
 
-class _FormattableMessage:
-    def __init__(self, message):
-        self.content = message.get("content")
-        self.role = message.get("role")
-        self.variables = _parse_format_fields(self.content)
+class _ContentFormatter:
+    def __init__(self, type, template=None):
+        if type == "completions":
+            if not template:
+                template = "{prompt}"
+            if not isinstance(template, str):
+                raise mlflow.MlflowException.invalid_parameter_value(
+                    f"Type {type(template)} is not a valid type for template of task {type}."
+                )
+
+            self.template = template
+            self.format_fn = self.format_prompt
+            self.variables = sorted(_parse_format_fields(self.template))
+        elif type == "chat.completions":
+            if not template:
+                template = [{"role": "user", "content": "{content}"}]
+            if not all(isinstance(message, Dict) for message in template):
+                raise mlflow.MlflowException.invalid_parameter_value(
+                    f"Type {type(template)} is not a valid type for template of task {type}."
+                )
+
+            self.template = template.copy()
+            self.format_fn = self.format_chat
+            self.variables = sorted(
+                set(
+                    itertools.chain.from_iterable(
+                        _parse_format_fields(message.get("content"))
+                        | _parse_format_fields(message.get("role"))
+                        for message in self.template
```

Comment:
potentially dumb question, but is there a guarantee that all messages within the template will have both "content" and "role" defined?

Reply:
Correct. All messages need to have a role and content. Now this is the tricky part. In the current implementation of OpenAI flavor, we assume that each data row in the inputs generates an entire conversation (sequence of messages) based on the template of messages. We build the `content` and `role` structure. I didn't change that behavior in this PR but I think we need to give users the ability to pass the history of messages with multiple input rows. For instance, someone should be able to use `predict` with columns `role` and `content`, each row representing one message in the conversation, and send that entire data frame as a single conversation.

Reply:
+1 I whole heartedly agree. That statefulness is really important for a 'chat bot' with context. I agree that that should probably be a follow-up PR :D

Reply:
Awesome. I will draft something and share it with you. Cause we need to find a way to introduce it without breaking former users (or allowing both use cases).